### ISSUE TITLE
performing the test phase on cpu or another gpu?

### ISSUE THREAD

**Issue by RSly** [NVIDIA/caffe#345]
hi,

In the latest nvcaffe 0.16, the test phase is performed on the same gpus as the training ones. This causes the training to be halted for the time that the testing takes. the problem is that when we performe on large test/validation sets, or when we want a high resolution of the validation results the training can be halted or delayed for many hours due to this sequential implementation.

is there already a way to perform the testing phase on a separate resource such as free cpus or another dedicated gpu?

I think this feature can bring a great advantage to achieving a faster workflow with caffe :)

**Comment by cliffwoolley** [comment]
You might be able to script something like this already.  If you are saving
snapshots at least as frequently as you're testing (and certainly I hope
you are), then you could have your trainer never test, only snapshot, and
you could have another Caffe instance entirely that loads in a snapshot
from disk, tests it against your validation set immediately (after 0
additional training steps, that is), and then exits right away.  Just set
up your script to invoke that tester every time a new snapshot appears on
the filesystem.

As for where to do this, I might actually recommend doing it on an entirely
different node if it's really all that heavyweight, such that your train
and test nets and their respective compute, memory, and I/O needs aren't
getting in each other's way.

-Cliff

**Comment by RSly** [comment]
thanks @cliffwoolley ! I was thinking of this also. I will try it out!

however, I still feel it is worthed to be part of the main nvcaffe in the future! 

ps. btw if you have already such a scrip it is greatly appreciated:)

**Comment by RSly** [comment]
FYI,

Related to the test phase, I noticed that when using multiple GPUs with caffe 0.16, the test phases are also running using multiple GPUs (before, the test phase always used one GPU).

This is also related to https://github.com/NVIDIA/DIGITS/issues/1604

**Comment by RSly** [comment]
I have a small related question: in the current implementation of the multi-gpu test scenario, how can a python data-layer know which GPU is running it? in other words, I need the python data layer to know which gpu is running it, so to make a deterministic test by giving to the gpu i the i-th portion of the test set

**Comment by drnikolaev** [comment]
@RSly well, test set distribution is deterministic. Please let me know if you observe opposite (if you do, make sure you use deterministic convolutional algos as well as max pooling algo).

**Comment by RSly** [comment]
hi @cliffwoolley @drnikolaev, it seems that nvcaffe ignores solver_mode: CPU. so I cant use the test-only script to run  in cpu mode. I have a test-only solver with these flags:
base_lr: 0
solver_mode:CPU
snapshot_after_train: false

I run the above solver (CPU) with tools/caffe train, and no --gpu flag, but nvcaffe still tries to use the busy gpus, and crashes.
I think these lines here ignore the solver_mode:CPU, and force the gpu to be set:
https://github.com/NVIDIA/caffe/blob/caffe-0.16/tools/caffe.cpp#L161-L171
and thus this if never happens?
https://github.com/NVIDIA/caffe/blob/caffe-0.16/tools/caffe.cpp#L183

is this a bug or maybe there is another way to tell the caffe/train to actually use CPU?
ps. if I remember, in older days, one could use /train ... --gpu=-1 to turn off gpu usage

**Comment by drnikolaev** [comment]
@RSly you need to rebuild with CPU_ONLY=ON to run on CPU.

**Comment by RSly** [comment]
yes, but that is not very efficient, we ended up having multiple sets of nv-caffe for test, train, etc... I think best could be to activate this CPU_ONLY=ON in the solver, or --gpu=-1 in the train to switch between CPU and GPU

* this aside, I strongly vote for being able to 'configure the test phase to perform on cpus or another free gpu'

**Comment by drnikolaev** [comment]
Got your point, we'll need to discuss&prioritize it. Marking and closing at this moment, thank you.
